NMS-13760: Split SNMP Property Extenders into multiple pages (#4086)
OpenNMS/opennms: docs/modules/operation/pages/performance-data-collection/snmp-property-extenders/pointer-like-index.adoc

> == Pointer-Like Index Property Extender

Reviewer: ```suggestion
= Pointer-Like Index Property Extender
```

The Pointer-Like Index property extender makes it possible to use the value of an attribute from the target MIB table as the index into the source MIB table.
Unlike the Index Split and Regex extenders, this extender class does not require the target and source MIB tables to share any index components.

> === When to Use the Pointer-Like Index Property Extender

Reviewer: ```suggestion
== When to Use the Pointer-Like Index Property Extender
```

            entPhysicalVendorType, and entPhysicalClass) to help an NMS
            identify and characterize the entry, and objects
            (entPhysicalContainedIn and entPhysicalParentRelPos) to help
            an NMS relate the particular entry to other entries in this
            table."
    INDEX   { entPhysicalIndex } <3>
    -- ...
----
<1> The `cpmCPUTotalTable` entry type is indexed on `cpmCPUTotalIndex`, which has no meaning outside this table
<2> The `cpmCPUTotalPhysicalIndex` column contains a value of `entPhysicalIndex` corresponding to the CPU referenced in a given row
<3> The `entPhysicalTable` entry type is indexed on `entPhysicalIndex` and provides many useful textual columns.

By treating `cpmCPUTotalPhysicalIndex` somewhat like a link:https://en.wikipedia.org/wiki/Pointer_(computer_programming)[pointer], it's possible to import string properties from the `entPhysicalTable` for use in the resource-label.

NOTE: Some combinations of Cisco hardware and software appear to use values of `cpmCPUTotalIndex` that are directly interchangeable with `entPhysicalIndex`.
This relationship does not hold across all product lines or software revisions.

> === Configuring the Pointer-Like Index Property Extender

Reviewer: ```suggestion
== Configuring the Pointer-Like Index Property Extender
```

The Pointer-Like Index property extender expects three parameters, all of which are required:

.Pointer-Like Index Property Extender Parameters
[options="header", cols="1,3"]
|===
| Name
| Description

| source-type
| The name of the `resourceType` associated with the source MIB table.

| source-attribute
| The alias name of the string property to import from the source MIB table.

| target-index-pointer-column
| The alias name of the column in the target MIB table whose value may be used as an index into the source MIB table.
|===

This example shows how to use `cpmCPUTotalPhysicalIndex` as a pointer-like index into the `entPhysicalTable`.
The target resource gains a pair of string properties, which we will call `cpmCPUTotalName` and `cpmCPUTotalDescr`.

[source, xml]
----

> <resourceType name="entPhysicalEntry" label="Physical Entity" resourceLabel="${entPhysicalName} (${entPhysicalDescr}))">
>    <persistenceSelectorStrategy class="org.opennms.netmgt.collection.support.PersistAllSelectorStrategy"/>
>    <storageStrategy class="org.opennms.netmgt.collection.support.IndexStorageStrategy"/>
> </resourceType>
> 
> <resourceType name="cpmCPUTotalEntry" label="Cisco CPU Total" resourceLabel="${cpmCPUTotalName} (${cpmCPUTotalDescr})">
>    <persistenceSelectorStrategy class="org.opennms.netmgt.collection.support.PersistAllSelectorStrategy" />
>    <storageStrategy class="org.opennms.netmgt.collection.support.IndexStorageStrategy" />
> </resourceType>
> 
> <groups>
>   <group name="entity-physical-table" ifType="all">
>     <mibObj oid=".1.3.6.1.2.1.47.1.1.1.1.2" instance="entPhysicalEntry" alias="entPhysicalDescr" type="string"/> <1>
>     <mibObj oid=".1.3.6.1.2.1.47.1.1.1.1.7" instance="entPhysicalEntry" alias="entPhysicalName" type="string"/> <2>
>   </group>
> 
>   <group name="cpm-cpu-total" ifType="all">
>     <mibObj oid=".1.3.6.1.4.1.9.9.109.1.1.1.1.2"  instance="cpmCPUTotalEntry" alias="cpmCPUTotalPhysicalIndex"  type="string" /> <3>
>     <mibObj oid=".1.3.6.1.4.1.9.9.109.1.1.1.1.8"  instance="cpmCPUTotalEntry" alias="cpmCPUTotal5minRev"        type="gauge" />
>     <property instance="cpmCPUTotalEntry" alias="cpmCPUTotalName" class-name="org.opennms.netmgt.collectd.PointerLikeIndexPropertyExtender"> <4>
>       <parameter key="source-type" value="entPhysicalEntry"/>
>       <parameter key="source-attribute" value="entPhysicalName"/> <5>
>       <parameter key="target-index-pointer-column" value="cpmCPUTotalPhysicalIndex"/>
>     </property>
>     <property instance="cpmCPUTotalEntry" alias="cpmCPUTotalDescr" class-name="org.opennms.netmgt.collectd.PointerLikeIndexPropertyExtender"> <6>
>       <parameter key="source-type" value="entPhysicalEntry"/>
>       <parameter key="source-attribute" value="entPhysicalDescr"/> <7>
>       <parameter key="target-index-pointer-column" value="cpmCPUTotalPhysicalIndex"/>
>     </property>
>   </group>
> </groups>

Reviewer: ```suggestion
<resourceType name="entPhysicalEntry" label="Physical Entity" resourceLabel="${entPhysicalName} (${entPhysicalDescr}))">
   <persistenceSelectorStrategy class="org.opennms.netmgt.collection.support.PersistAllSelectorStrategy"/>
   <storageStrategy class="org.opennms.netmgt.collection.support.IndexStorageStrategy"/>
</resourceType>

<resourceType name="cpmCPUTotalEntry" label="Cisco CPU Total" resourceLabel="${cpmCPUTotalName} (${cpmCPUTotalDescr})">
   <persistenceSelectorStrategy class="org.opennms.netmgt.collection.support.PersistAllSelectorStrategy" />
   <storageStrategy class="org.opennms.netmgt.collection.support.IndexStorageStrategy" />
</resourceType>

<groups>
  <group name="entity-physical-table" ifType="all">
    <mibObj oid=".1.3.6.1.2.1.47.1.1.1.1.2" instance="entPhysicalEntry" alias="entPhysicalDescr" type="string"/> <1>
    <mibObj oid=".1.3.6.1.2.1.47.1.1.1.1.7" instance="entPhysicalEntry" alias="entPhysicalName" type="string"/> <2>
  </group>

  <group name="cpm-cpu-total" ifType="all">
    <mibObj oid=".1.3.6.1.4.1.9.9.109.1.1.1.1.2"  instance="cpmCPUTotalEntry" alias="cpmCPUTotalPhysicalIndex"  type="string" /> <3>
    <mibObj oid=".1.3.6.1.4.1.9.9.109.1.1.1.1.8"  instance="cpmCPUTotalEntry" alias="cpmCPUTotal5minRev"        type="gauge" />
    <property instance="cpmCPUTotalEntry" alias="cpmCPUTotalName" class-name="org.opennms.netmgt.collectd.PointerLikeIndexPropertyExtender"> <4>
      <parameter key="source-type" value="entPhysicalEntry"/>
      <parameter key="source-attribute" value="entPhysicalName"/> <5>
      <parameter key="target-index-pointer-column" value="cpmCPUTotalPhysicalIndex"/>
    </property>
    <property instance="cpmCPUTotalEntry" alias="cpmCPUTotalDescr" class-name="org.opennms.netmgt.collectd.PointerLikeIndexPropertyExtender"> <6>
      <parameter key="source-type" value="entPhysicalEntry"/>
      <parameter key="source-attribute" value="entPhysicalDescr"/> <7>
      <parameter key="target-index-pointer-column" value="cpmCPUTotalPhysicalIndex"/>
    </property>
  </group>
</groups>
```

Author: @mmahacek What's the change here?
GH does not like it:
![image](https://user-images.githubusercontent.com/4995222/150863456-571592bd-6ce7-45ad-88e5-318be92f1023.png)

Reviewer: Sorry, @mfuhrmann, I had copied the wrong block from my local system.  The numbers need to be adjusted to line up with the updated footnotes.

----
<1>, <2> First we collect `entPhysicalDescr` and `entPhysicalName` in the source group, which uses a resource type associated with the `entPhysicalTable`.
<3> Then we collect the pointer-like `cpmCPUTotalPhysicalIndex` in the target group, whose resource type is associated with the `cpmCPUTotalTable`.

> <4> We derive `cpmCPUTotalName` in the target group, telling the extender to use the pointer-like property's value as an index into the source table, and specifying that we want to "pull over" the source attribute `entPhysicalName`. <5>

Author: ```suggestion
<4> We derive `cpmCPUTotalName` in the target group, telling the extender to use the pointer-like property's value as an index into the source table, and specifying that we want to use the source attribute `entPhysicalName`. <5>
```

> <1>, <2> First we collect `entPhysicalDescr` and `entPhysicalName` in the source group, which uses a resource type associated with the `entPhysicalTable`.
> <3> Then we collect the pointer-like `cpmCPUTotalPhysicalIndex` in the target group, whose resource type is associated with the `cpmCPUTotalTable`.
> <4> We derive `cpmCPUTotalName` in the target group, telling the extender to use the pointer-like property's value as an index into the source table, and specifying that we want to use the source attribute `entPhysicalName`. <5>
> <6> Deriving `cpmCPUTotalDescr` is almost identical, except that this time we use the value of `entPhysicalDescr`. <7>

Reviewer: Footnote numbering is off.
```suggestion
<1> First we collect `entPhysicalDescr` and `entPhysicalName` in the source group, which uses a resource type associated with the `entPhysicalTable`.
<2> Then we collect the pointer-like `cpmCPUTotalPhysicalIndex` in the target group, whose resource type is associated with the `cpmCPUTotalTable`.
<3> We derive `cpmCPUTotalName` in the target group, telling the extender to use the pointer-like property's value as an index into the source table, and specifying that we want to use the source attribute `entPhysicalName`.
<4> Deriving `cpmCPUTotalDescr` is almost identical, except that this time we use the value of `entPhysicalDescr`.
```

> <6> Deriving `cpmCPUTotalDescr` is almost identical, except that this time we pull over the value of `entPhysicalDescr`. <7>

Author: ```suggestion
<6> Deriving `cpmCPUTotalDescr` is almost identical, except that this time we use the value of `entPhysicalDescr`. <7>
```

> [source, xml]
> ----
> <resourceType name="entPhysicalEntry" label="Physical Entity" resourceLabel="${entPhysicalName} (${entPhysicalDescr}))">
>    <persistenceSelectorStrategy class="org.opennms.netmgt.collection.support.PersistAllSelectorStrategy"/>
>    <storageStrategy class="org.opennms.netmgt.collection.support.IndexStorageStrategy"/>
> </resourceType>
> 
> <resourceType name="cpmCPUTotalEntry" label="Cisco CPU Total" resourceLabel="${cpmCPUTotalName} (${cpmCPUTotalDescr})">
>    <persistenceSelectorStrategy class="org.opennms.netmgt.collection.support.PersistAllSelectorStrategy" />
>    <storageStrategy class="org.opennms.netmgt.collection.support.IndexStorageStrategy" />
> </resourceType>
> 
> <groups>
>   <group name="entity-physical-table" ifType="all">
>     <mibObj oid=".1.3.6.1.2.1.47.1.1.1.1.2" instance="entPhysicalEntry" alias="entPhysicalDescr" type="string"/> <1>
>     <mibObj oid=".1.3.6.1.2.1.47.1.1.1.1.7" instance="entPhysicalEntry" alias="entPhysicalName" type="string"/> <2>
>   </group>
> 
>   <group name="cpm-cpu-total" ifType="all">
>     <mibObj oid=".1.3.6.1.4.1.9.9.109.1.1.1.1.2"  instance="cpmCPUTotalEntry" alias="cpmCPUTotalPhysicalIndex"  type="string" /> <3>
>     <mibObj oid=".1.3.6.1.4.1.9.9.109.1.1.1.1.8"  instance="cpmCPUTotalEntry" alias="cpmCPUTotal5minRev"        type="gauge" />
>     <property instance="cpmCPUTotalEntry" alias="cpmCPUTotalName" class-name="org.opennms.netmgt.collectd.PointerLikeIndexPropertyExtender"> <4>
>       <parameter key="source-type" value="entPhysicalEntry"/>
>       <parameter key="source-attribute" value="entPhysicalName"/> <5>
>       <parameter key="target-index-pointer-column" value="cpmCPUTotalPhysicalIndex"/>
>     </property>
>     <property instance="cpmCPUTotalEntry" alias="cpmCPUTotalDescr" class-name="org.opennms.netmgt.collectd.PointerLikeIndexPropertyExtender"> <6>
>       <parameter key="source-type" value="entPhysicalEntry"/>
>       <parameter key="source-attribute" value="entPhysicalDescr"/> <7>
>       <parameter key="target-index-pointer-column" value="cpmCPUTotalPhysicalIndex"/>
>     </property>
>   </group>
> </groups>
> ----
> <1>, <2> First we collect `entPhysicalDescr` and `entPhysicalName` in the source group, which uses a resource-type associated with the `entPhysicalTable`
> <3> Then we collect the pointer-like `cpmCPUTotalPhysicalIndex` in the target group, whose resource-type is associated with the `cpmCPUTotalTable`
> <4> We derive `cpmCPUTotalName` in the target group telling the extender to use the pointer-like property's value as an index into the source table, and specify that we want to "pull over" the source attribute `entPhysicalName` <5>
> <6> Deriving `cpmCPUTotalDescr` is almost identical, except that this time we are pulling over the value of `entPhysicalDescr` <7>

Reviewer: The footnote numbering was off.
```suggestion
[source, xml]
----
<resourceType name="entPhysicalEntry" label="Physical Entity" resourceLabel="${entPhysicalName} (${entPhysicalDescr}))">
   <persistenceSelectorStrategy class="org.opennms.netmgt.collection.support.PersistAllSelectorStrategy"/>
   <storageStrategy class="org.opennms.netmgt.collection.support.IndexStorageStrategy"/>
</resourceType>

<resourceType name="cpmCPUTotalEntry" label="Cisco CPU Total" resourceLabel="${cpmCPUTotalName} (${cpmCPUTotalDescr})">
   <persistenceSelectorStrategy class="org.opennms.netmgt.collection.support.PersistAllSelectorStrategy" />
   <storageStrategy class="org.opennms.netmgt.collection.support.IndexStorageStrategy" />
</resourceType>

<groups>
  <group name="entity-physical-table" ifType="all">
    <mibObj oid=".1.3.6.1.2.1.47.1.1.1.1.2" instance="entPhysicalEntry" alias="entPhysicalDescr" type="string"/> <1>
    <mibObj oid=".1.3.6.1.2.1.47.1.1.1.1.7" instance="entPhysicalEntry" alias="entPhysicalName" type="string"/> <2>
  </group>

  <group name="cpm-cpu-total" ifType="all">
    <mibObj oid=".1.3.6.1.4.1.9.9.109.1.1.1.1.2"  instance="cpmCPUTotalEntry" alias="cpmCPUTotalPhysicalIndex"  type="string" /> <3>
    <mibObj oid=".1.3.6.1.4.1.9.9.109.1.1.1.1.8"  instance="cpmCPUTotalEntry" alias="cpmCPUTotal5minRev"        type="gauge" />
    <property instance="cpmCPUTotalEntry" alias="cpmCPUTotalName" class-name="org.opennms.netmgt.collectd.PointerLikeIndexPropertyExtender"> <4>
      <parameter key="source-type" value="entPhysicalEntry"/>
      <parameter key="source-attribute" value="entPhysicalName"/>
      <parameter key="target-index-pointer-column" value="cpmCPUTotalPhysicalIndex"/>
    </property>
    <property instance="cpmCPUTotalEntry" alias="cpmCPUTotalDescr" class-name="org.opennms.netmgt.collectd.PointerLikeIndexPropertyExtender"> <5>
      <parameter key="source-type" value="entPhysicalEntry"/>
      <parameter key="source-attribute" value="entPhysicalDescr"/>
      <parameter key="target-index-pointer-column" value="cpmCPUTotalPhysicalIndex"/>
    </property>
  </group>
</groups>
----
<1> First we collect `entPhysicalDescr` and `entPhysicalName` in the source group, which uses a resource-type associated with the `entPhysicalTable`.
<2> See above.
<3> Then we collect the pointer-like `cpmCPUTotalPhysicalIndex` in the target group, whose resource-type is associated with the `cpmCPUTotalTable`.
<4> We derive `cpmCPUTotalName` in the target group telling the extender to use the pointer-like property's value as an index into the source table, and specify that we want to "pull over" the source attribute `entPhysicalName`.
<5> Deriving `cpmCPUTotalDescr` is almost identical, except that this time we are pulling over the value of `entPhysicalDescr`.
```

> <6> Deriving `cpmCPUTotalDescr` is almost identical, except that this time we are pulling over the value of `entPhysicalDescr` <7>

Author: ```suggestion
<6> Deriving `cpmCPUTotalDescr` is almost identical, except that this time we pul over the value of `entPhysicalDescr` <7>
```
